extension/arquillian-drone-browserstack-extension/src/main/java/org/arquillian/drone/browserstack/extension/local/BrowserStackLocalRunner.java

```diff
@@ -0,0 +1,257 @@
+        if (process != null) {
+            return;
+        }
+        if (isEmpty(localBinary)) {
+            if (!browserStackLocalFile.exists()) {
+                prepareBrowserStackLocal();
+            }
+            runTheBinary(browserStackLocalFile, accessKey, additionalArgs);
+
+        } else {
+            runTheBinary(new File(localBinary), accessKey, additionalArgs);
+        }
+    }
+
+    /**
+     * Runs BrowserStackLocal binary. In case that the binary has been already run, then does nothing.
+     *
+     * @param binaryFile     A binary file to be run
+     * @param accessKey      An accessKey the binary should be ran with
+     * @param additionalArgs additional arguments
+     */
+    private void runTheBinary(File binaryFile, String accessKey, String additionalArgs) {
+        List<String> args = new ArrayList<String>();
+        args.add(binaryFile.getAbsolutePath());
+        args.add(accessKey);
+        if (!isEmpty(additionalArgs)) {
+            args.addAll(Arrays.asList(additionalArgs.split(" ")));
+        }
+        ProcessBuilder processBuilder = new ProcessBuilder().command(args);
+
+        try {
+            process = processBuilder.start();
+
+            final Reader reader = new Reader();
+            reader.start();
+            Runtime.getRuntime().addShutdownHook(new CloseChildProcess());
+            countDownLatch.await(20, TimeUnit.SECONDS);
+
+        } catch (UnsupportedEncodingException e) {
+            e.printStackTrace();
+        } catch (IOException e) {
+            e.printStackTrace();
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+    }
+
+    /**
+     * Prepares the BrowserStackLocal binary. Creates the directory target/browserstacklocal; downloads a zip file
+     * containing the binary; extracts it into the created directory and marks the binary as executable.
+     */
+    private void prepareBrowserStackLocal() {
+        String platformBinaryNameUrl = getPlatformBinaryNameUrl();
+        File browserStackLocalZipFile =
+            new File(browserStackLocalDirectory.getPath() + File.separator + platformBinaryNameUrl);
+        String url = basicUrl + platformBinaryNameUrl;
+
+        log.info("Creating directory: " + browserStackLocalDirectory);
+        browserStackLocalDirectory.mkdir();
+
+        log.info("downloading zip file from: " + url + " to " + browserStackLocalZipFile.getPath());
+        Spacelift.task(DownloadTool.class).from(url).to(browserStackLocalZipFile.getPath()).execute().await();
+
+        log.info("extracting zip file: " + browserStackLocalZipFile + " to " + browserStackLocalDirectory.getPath());
+        Spacelift.task(browserStackLocalZipFile, UnzipTool.class).toDir(browserStackLocalDirectory.getPath()).execute()
+            .await();
+
+        log.info("marking binary file: " + browserStackLocalFile.getPath() + " as executable");
+        browserStackLocalFile.setExecutable(true);
```

Comment:
Yes, but it is an exception with a meaningful message, so the user should understand where the problem is.